core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2670,6 +2573,112 @@ object Log {
   private def isLogFile(file: File): Boolean =
     file.getPath.endsWith(LogFileSuffix)
 
+  // Rebuild producer state until lastOffset. This method may be called from the recovery code path, and thus must be
+  // free of all side-effects, i.e. it must not update any log-specific state.
+  private def rebuildProducerState(log: Log,
```

Comment:
This seems too large as a companion method. I am also not sure that we should pass along Log to this method since it exposes all internal states in Log, which make it harder to track usage.

Reply:
That is true. Do you feel the original approach in f0466b2420dc500112b4839678564996c0677d2e was better, where, we didn't move methods into a companion object? I'd be happy to revert to that.

Reply:
We can probably keep the two other two simpler methods (without Log as the input) as the companion methods.